packages/opencensus-web-instrumentation-zone/src/perf-resource-timing-selector.ts

```diff
@@ -0,0 +1,183 @@
+    // Compare every performance entry with the perfomance entries in front of
+    // it. This is possible as the entries are sorted by the startTime. That
+    // way we to avoid comparing twice the entries and taking the wrong order.
+    for (let j = i + 1; j < filteredSortedPerfEntries.length; j++) {
+      const entryJ = filteredSortedPerfEntries[j];
+      if (isPossibleCorsPair(entryI, entryJ)) {
+        // As the entries are not overlapping, that means those timings
+        // are possible perfomance timings related to the XHR.
+        possiblePerfEntries.push({
+          corsPreFlightRequest: entryI,
+          mainRequest: entryJ,
+        });
+      }
+    }
+  }
+  return possiblePerfEntries;
+}
+
+/**
+ * Pick the best performance resource timing for the XHR: Using the possible
+ * performance resource timing entries from previous step, the best entry will
+ * be the one with the minimum gap to the span start/end timings.
+ * The performance resource timing entry with the minimum gap to the span
+ * start/end timings points out that entry is the best fit for the span.
+ *
+ * @param perfEntries
+ * @param span
+ */
+function getBestPerfResourceTiming(
+  perfEntries: XhrPerformanceResourceTiming[],
+  span: Span
+): XhrPerformanceResourceTiming | undefined {
+  let minimumGapToSpan = Number.MAX_VALUE;
+  let bestPerfEntry: XhrPerformanceResourceTiming | undefined;
+  for (const perfEntry of perfEntries) {
+    // If the current entry has cors preflight data use its `startTime` to
+    // calculate the gap to the span.
+    const perfEntryStartTime = perfEntry.corsPreFlightRequest
+      ? perfEntry.corsPreFlightRequest.startTime
+      : perfEntry.mainRequest.startTime;
+    const gapToSpan =
+      span.endPerfTime -
+      perfEntry.mainRequest.responseEnd +
+      (perfEntryStartTime - span.startPerfTime);
+
+    // If there is a new minimum gap to the span, update the minimum and pick
+    // the current performance entry as the best at this point.
+    if (gapToSpan < minimumGapToSpan) {
+      minimumGapToSpan = gapToSpan;
+      bestPerfEntry = perfEntry;
+    }
+  }
+  return bestPerfEntry;
+}
+
+/**
+ * A Performance entry is part of a XHR if entry has not been assigned
+ * previously to another XHR and the URL is the same as the XHR and the
+ * entry's start/end times are within the span's start/end times.
+ */
+function isPerfEntryPartOfXhr(
+  entry: PerformanceResourceTiming,
+  xhrUrl: string,
+  span: Span
+): boolean {
+  return (
+    !alreadyAssignedPerfEntries.has(entry) &&
+    entry.name === xhrUrl &&
+    entry.startTime >= span.startPerfTime &&
+    entry.responseEnd <= span.endPerfTime
+  );
+}
+
+/**
+ * A possible CORS pair is defined when the entries does not overlap in their
+ * start/end times.
+ */
+function isPossibleCorsPair(
+  entry1: PerformanceResourceTiming,
+  entry2: PerformanceResourceTiming
+): boolean {
+  // To determine if the entries overlap, the minimum responseEnd should be
```

Comment:
Could you simplify this logic if you made `entry1` be called `maybePreflight` and `entry2` be called `maybeMainRequset`? The assumption there is that the `maybePreflight` start time is definitely less than the `maybeMainRequest` start time, because of the sorting done by `getEntriesByType`